Force interpretation of Swagger UI js as UTF-8

abutch3r opened the pull request:
By not explicitly setting the charset on the Javascript it can potentially be interpreted as ISO-8859-1 causing `&nbsp;` to be seen as `Â ` (space included) due to the fact that they have the characters have matching byte code.

Setting the charset on the script in the html files forces the browser to view that file in the stated character set.

Apply to all html files for Swagger UI

Modify UI Basic Selenium tests to check the value of the text for a required parameter matches the expected UTF-8 value for it. if it contains the alternative it will fail

Setting the content-type header to include `charset=UTF-8` would be sufficient. however by updating the HTML to include the interpretation of the JS in UTF-8.

Fixes #26775
Fixes #26655

abutch3r committed:
Force interpretation of Swagger UI js as UTF-8

By not setting the charset on the Javascript it can potentially be interpreted as ISO-8859-1 causing `&nbsp;` to be seen as `Â ` (space included) due to the fact that they have the characters have matching byte code.

Setting the charset on the script in the html files forces the browser to view that file in the stated character set.

Apply to all html files for Swagger

Modify UI Basic Selenium tests to check the value of the text for a required parameter matches the expected UTF-8 value for it. if it contains the alternative it will fail

abutch3r commented:
#build

abutch3r committed:
Set OpenAPI UI html to use UTF-8 charset

As the Servlet Filter manipulates the html before it is sent to browser. It is adding the Servlet default charset which is UTF-8859-1 onto the content-type. By setting the header on the response to be the value with UTF-8 it is forced to be intrepreted as such.

abutch3r commented:
#libby
#build

Azquelt commented on dev/com.ibm.ws.openapi.ui/swagger-ui/dist/index.html:
```diff
@@ -5,7 +5,8 @@
     <link rel="icon" type="image/png" href="./favicon-32x32.png" sizes="32x32" />
     <link rel="icon" type="image/png" href="./favicon-16x16.png" sizes="16x16" />
-    <script defer src="liberty-swagger-ui-bundle-f46b8f492bc9c6c9d41c.js"></script><link href="liberty-swagger-ui-a44cef3790da8bb191bd.css" rel="stylesheet">
+    <script defer src="liberty-swagger-ui-bundle-f46b8f492bc9c6c9d41c.js" charset="UTF-8"></script>
+    <link href="liberty-swagger-ui-a44cef3790da8bb191bd.css" rel="stylesheet">
```
These files are generated by the NPM build. You can't edit them by hand.

abutch3r replied:
Have removed forcesetting here - would need to see if there was anything in the webpack steps where we could add for extra security as while potentially doable elsewhere it would cause a difference between `/dist` contents and what the server was seeing

Azquelt commented on dev/com.ibm.ws.microprofile.openapi/src/com/ibm/ws/microprofile/openapi/servlet/filter/OpenAPIEndpointFilter.java:
```diff
@@ -52,15 +53,16 @@ public void doFilter(ServletRequest req, ServletResponse resp, FilterChain chain
                 }
-                resp.setContentLength(content.length());
-                resp.getWriter().write(content);
+                httpServletResp.setHeader("Content-Type",httpServletResp.getContentType()+";charset=UTF-8");
```
I'm not convinced this is the right way to do this. Should we be calling `httpServletResp.setCharacterEncoding()` instead?

abutch3r replied:
Yes, as it preserves the other part of the Content-Type for us, have changed to use it instead of constructing the header ourselves

Azquelt commented on dev/com.ibm.ws.microprofile.openapi/src/com/ibm/ws/microprofile/openapi/servlet/filter/OpenAPIEndpointFilter.java:
```diff
@@ -52,15 +53,16 @@ public void doFilter(ServletRequest req, ServletResponse resp, FilterChain chain
             //Modify Response
-            if (resp.getContentType() != null && resp.getContentType().contains("text/html")) {
+            if (httpServletResp.getContentType() != null && httpServletResp.getContentType().contains("text/html")) {
                 String content = wrapper.getContentAsString();
                 // in case we have had a issue creating the tracker earlier - if so leave content as-is
                 if(openAPIEndpointTracker != null) {
                     // replace the default URL for the document endpoint with the value that is being used
                     content = content.replaceAll("/openapi", openAPIEndpointTracker.getService().getOpenAPIDocUrl());
                 }
-                resp.setContentLength(content.length());
-                resp.getWriter().write(content);
+                httpServletResp.setHeader("Content-Type",httpServletResp.getContentType()+";charset=UTF-8");
+                httpServletResp.setContentLength(content.getBytes(StandardCharsets.UTF_8).length);
+                httpServletResp.getWriter().write(content);
             }
         }
```
I have a second concern about this as well.

The HTML is being read from a file, presumably as bytes. In this filter, we're then asking for the content as a string.

Given the problems we have had have been down to the servlet container defaulting to ISO-8859-1, is the servlet container interpreting the bytes from the file as ISO-8859-1 rather than UTF-8?

This may not cause problems with our specific files if they're entirely ascii, but it does suggest we're doing something wrong here that may fail in the future.

Azquelt replied:
Possibly the right thing to do is to get and set bytes and do any encoding ourselves, relying on the fact that we know the source files are UTF-8.

Possibly we also want a better check that we're only operating on the index.html file.

abutch3r replied:
Have changed getContentAsString with getContentAsBytes and now construct the String with that using UTF-8 to reduce any potential ISO issues.

abutch3r committed:
improve encoding handling in servlet filter

remove html changes as cannot currently be maintained when the javascript bundle line is added to the HTML

abutch3r commented:
#libby
#build

abutch3r commented:
#libby
#build

abutch3r commented:
#libby
